// mjml/lib/components/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _MJMLColumnElement = require('./decorators/MJMLColumnElement');

var _MJMLColumnElement2 = _interopRequireDefault(_MJMLColumnElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Displays a customizable button
 */
var Button = (_dec = (0, _MJMLColumnElement2.default)({
  tagName: 'mj-button',
  content: '',
  attributes: {
    'align': 'center',
    'background-color': '#414141',
    'border-radius': '3px',
    'border': 'none',
    'color': '#ffffff',
    'font-family': 'Ubuntu, Helvetica, Arial, sans-serif',
    'font-size': '13px',
    'font-weight': 'normal',
    'href': '',
    'padding': '10px 25px',
    'text-decoration': 'none',
    'vertical-align': 'middle'
  }
}), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Button);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Button)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Button, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        td: {
          background: mjAttribute('background-color'),
          borderRadius: mjAttribute('border-radius'),
          color: mjAttribute('color'),
          cursor: 'auto',
          fontStyle: mjAttribute('font-style')
        },
        table: {
          border: mjAttribute('border'),
          borderRadius: mjAttribute('border-radius')
        },
        a: {
          background: mjAttribute('background-color'),
          border: '1px solid ' + mjAttribute('background-color'),
          borderRadius: mjAttribute('border-radius'),
          color: mjAttribute('color'),
          fontFamily: mjAttribute('font-family'),
          fontSize: mjAttribute('font-size'),
          fontStyle: mjAttribute('font-style'),
          fontWeight: mjAttribute('font-weight'),
          padding: mjAttribute('padding'),
          textDecoration: mjAttribute('text-decoration')
        }
      });
    }
  }, {
    key: 'renderButton',
    value: function renderButton() {
      var _props = this.props;
      var mjContent = _props.mjContent;
      var mjAttribute = _props.mjAttribute;


      return _react2.default.createElement('a', {
        dangerouslySetInnerHTML: { __html: mjContent() },
        href: mjAttribute('href'),
        style: this.styles.a,
        target: '_blank' });
    }
  }, {
    key: 'render',
    value: function render() {
      var mjAttribute = this.props.mjAttribute;


      return _react2.default.createElement(
        'table',
        {
          cellPadding: '0',
          cellSpacing: '0',
          'data-legacy-align': mjAttribute('align'),
          'data-legacy-border': '0',
          style: this.styles.table },
        _react2.default.createElement(
          'tbody',
          null,
          _react2.default.createElement(
            'tr',
            null,
            _react2.default.createElement(
              'td',
              {
                'data-legacy-align': 'center',
                'data-legacy-bgcolor': mjAttribute('background-color'),
                'data-legacy-valign': mjAttribute('vertical-align'),
                style: this.styles.td },
              this.renderButton()
            )
          )
        )
      );
    }
  }]);

  return Button;
}(_react.Component), _class2.baseStyles = {
  a: {
    display: 'inline-block',
    textDecoration: 'none'
  }
}, _temp2)) || _class);
exports.default = Button;
