// mjml/lib/components/Column.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _mjAttribute = require('../helpers/mjAttribute');

var _MJMLElement = require('./decorators/MJMLElement');

var _MJMLElement2 = _interopRequireDefault(_MJMLElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Columns are the basic containers for your content. They must be located under mj-section tags in order to be considered by the engine
 */
var Column = (_dec = (0, _MJMLElement2.default)({
  tagName: 'mj-column'
}), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Column, _Component);

  function Column() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Column);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Column)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Column, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        div: {
          display: 'inline-block',
          verticalAlign: mjAttribute('vertical-align'),
          fontSize: '13',
          textAlign: 'left',
          width: '100%'
        },
        table: {
          verticalAlign: mjAttribute('vertical-align'),
          background: mjAttribute('background-color')
        }
      });
    }
  }, {
    key: 'getColumnClass',
    value: function getColumnClass() {
      var _props = this.props;
      var mjAttribute = _props.mjAttribute;
      var sibling = _props.sibling;

      var width = mjAttribute('width');

      if (width == undefined) {
        return 'mj-column-per-' + parseInt(100 / sibling);
      }

      var _widthParser = (0, _mjAttribute.widthParser)(width);

      var parsedWidth = _widthParser.width;
      var unit = _widthParser.unit;


      switch (unit) {
        case '%':
          return 'mj-column-per-' + parsedWidth;

        case 'px':
        default:
          return 'mj-column-px-' + parsedWidth;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props;
      var mjAttribute = _props2.mjAttribute;
      var children = _props2.children;
      var sibling = _props2.sibling;

      var width = mjAttribute('width') || 100 / sibling;
      var mjColumnClass = this.getColumnClass();

      return _react2.default.createElement(
        'div',
        {
          'aria-labelledby': mjColumnClass,
          className: mjColumnClass,
          'data-column-width': parseInt(width),
          'data-vertical-align': this.styles.div.verticalAlign,
          style: this.styles.div },
        _react2.default.createElement(
          'table',
          {
            cellPadding: '0',
            cellSpacing: '0',
            'data-legacy-background': mjAttribute('background'),
            'data-legacy-border': '0',
            style: this.styles.table,
            width: '100%' },
          _react2.default.createElement(
            'tbody',
            null,
            children
          )
        )
      );
    }
  }]);

  return Column;
}(_react.Component), _class2.baseStyles = {
  div: {
    verticalAlign: 'top'
  }
}, _temp2)) || _class);
exports.default = Column;
