// mjml/lib/components/Divider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _MJMLColumnElement = require('./decorators/MJMLColumnElement');

var _MJMLColumnElement2 = _interopRequireDefault(_MJMLColumnElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _mjAttribute = require('../helpers/mjAttribute');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Displays a customizable divider
 */
var Divider = (_dec = (0, _MJMLColumnElement2.default)({
  tagName: 'mj-divider',
  attributes: {
    'border-color': '#000000',
    'border-style': 'solid',
    'border-width': '4px',
    'padding': '10px 25px',
    'width': '100%'
  }
}), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Divider, _Component);

  function Divider() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Divider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Divider)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Divider, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        p: {
          borderTop: mjAttribute('border-width') + ' ' + mjAttribute('border-style') + ' ' + mjAttribute('border-color'),
          width: mjAttribute('width')
        }
      });
    }
  }, {
    key: 'outlookWidth',
    value: function outlookWidth() {
      var mjAttribute = this.props.mjAttribute;

      var parentWidth = parseInt(mjAttribute('parentWidth'));

      var _widthParser = (0, _mjAttribute.widthParser)(mjAttribute('width'));

      var width = _widthParser.width;
      var unit = _widthParser.unit;


      switch (unit) {
        case '%':
          {
            return parentWidth * width / 100;
          }
        default:
          {
            return width;
          }
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement('p', {
        className: 'mj-divider-outlook',
        'data-divider-width': this.outlookWidth(),
        style: this.styles.p });
    }
  }]);

  return Divider;
}(_react.Component), _class2.baseStyles = {
  p: {
    fontSize: '1px',
    margin: '0 auto'
  }
}, _temp2)) || _class);
exports.default = Divider;
