// mjml/lib/components/Image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _min2 = require('lodash/min');

var _min3 = _interopRequireDefault(_min2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _MJMLColumnElement = require('./decorators/MJMLColumnElement');

var _MJMLColumnElement2 = _interopRequireDefault(_MJMLColumnElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Displays an image to your email. It is mostly similar to the HTML img tag
 */
var Image = (_dec = (0, _MJMLColumnElement2.default)({
  tagName: 'mj-image',
  attributes: {
    'height': 'auto',
    'padding': '10px 25px',
    'align': 'center',
    'alt': '',
    'border': 'none',
    'href': '',
    'src': '',
    'target': '_blank'
  }
}), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Image, _Component);

  function Image() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Image);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Image)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Image, [{
    key: 'getContentWidth',
    value: function getContentWidth() {
      var _props = this.props;
      var mjAttribute = _props.mjAttribute;
      var getPadding = _props.getPadding;

      var parentWidth = mjAttribute('parentWidth');

      var width = (0, _min3.default)([parseInt(mjAttribute('width')), parseInt(parentWidth)]);

      var paddingRight = getPadding('right');
      var paddingLeft = getPadding('left');
      var widthOverflow = paddingLeft + paddingRight + width - parseInt(parentWidth);

      return widthOverflow > 0 ? width - widthOverflow : width;
    }
  }, {
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        td: {
          width: this.getContentWidth()
        },
        img: {
          border: mjAttribute('border'),
          height: mjAttribute('height')
        }
      });
    }
  }, {
    key: 'renderImage',
    value: function renderImage() {
      var mjAttribute = this.props.mjAttribute;


      var img = _react2.default.createElement('img', {
        alt: mjAttribute('alt'),
        border: '0',
        src: mjAttribute('src'),
        style: this.styles.img,
        width: this.getContentWidth(),
        height: this.styles.img.height });

      if (mjAttribute('href') != '') {
        return _react2.default.createElement(
          'a',
          { href: mjAttribute('href'), target: mjAttribute('target') },
          img
        );
      } else {
        return img;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var mjAttribute = this.props.mjAttribute;


      return _react2.default.createElement(
        'table',
        {
          cellPadding: '0',
          cellSpacing: '0',
          'data-legacy-align': mjAttribute('align'),
          'data-legacy-border': '0',
          style: this.styles.table },
        _react2.default.createElement(
          'tbody',
          null,
          _react2.default.createElement(
            'tr',
            null,
            _react2.default.createElement(
              'td',
              { style: this.styles.td },
              this.renderImage()
            )
          )
        )
      );
    }
  }]);

  return Image;
}(_react.Component), _class2.baseStyles = {
  table: {
    borderCollapse: 'collapse',
    borderSpacing: '0'
  },
  img: {
    border: 'none',
    display: 'block',
    outline: 'none',
    textDecoration: 'none',
    width: '100%'
  }
}, _temp2)) || _class);
exports.default = Image;
