// mjml/lib/components/Location.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class;

var _MJMLColumnElement = require('./decorators/MJMLColumnElement');

var _MJMLColumnElement2 = _interopRequireDefault(_MJMLColumnElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Image = require('./Image');

var _Image2 = _interopRequireDefault(_Image);

var _Text = require('./Text');

var _Text2 = _interopRequireDefault(_Text);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Location = (_dec = (0, _MJMLColumnElement2.default)({
  tagName: 'mj-location',
  attributes: {
    'color': '#3aa7ed',
    'font-family': 'Roboto, sans-serif',
    'font-size': '18px',
    'font-weight': 500,
    'padding': '10px 25px',
    'img-src': 'http://i.imgur.com/DPCJHhy.png'
  }
}), _dec(_class = function (_Component) {
  _inherits(Location, _Component);

  function Location() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Location);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Location)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Location, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        text: {
          color: mjAttribute('color'),
          textDecoration: 'none'
        }
      });
    }
  }, {
    key: 'getAttributes',
    value: function getAttributes() {
      var mjAttribute = this.props.mjAttribute;


      return {
        text: {
          'font-family': mjAttribute('font-family'),
          'font-size': mjAttribute('font-size'),
          'font-weight': mjAttribute('font-weight'),
          'text-decoration': mjAttribute('text-decoration'),
          'padding': 0
        },
        img: {
          'padding': 0,
          'src': mjAttribute('img-src')
        }
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var mjAttribute = this.props.mjAttribute;

      var attrs = this.getAttributes();

      var address = 'http://maps.google.com/maps?q=' + encodeURIComponent(mjAttribute('address'));
      var text = mjAttribute('text') || mjAttribute('address');

      return _react2.default.createElement(
        'table',
        { width: '100%' },
        _react2.default.createElement(
          'tbody',
          null,
          _react2.default.createElement(_Image2.default, _extends({}, attrs.img, {
            href: address })),
          _react2.default.createElement(
            _Text2.default,
            _extends({}, attrs.text, {
              align: 'center' }),
            _react2.default.createElement(
              'a',
              { style: this.styles.text, href: address, target: '_blank' },
              text
            )
          )
        )
      );
    }
  }]);

  return Location;
}(_react.Component)) || _class);
exports.default = Location;
