// mjml/lib/components/Raw.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class;

var _MJMLElement = require('./decorators/MJMLElement');

var _MJMLElement2 = _interopRequireDefault(_MJMLElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Raw = (_dec = (0, _MJMLElement2.default)({
  tagName: 'mj-raw'
}), _dec(_class = function (_Component) {
  _inherits(Raw, _Component);

  function Raw() {
    _classCallCheck(this, Raw);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(Raw).apply(this, arguments));
  }

  _createClass(Raw, [{
    key: 'getTagName',
    value: function getTagName() {
      var parentMjml = this.props.parentMjml;

      var tagName = void 0;

      switch (parentMjml.get('tagName')) {
        case 'mj-column':
          tagName = 'tr';
          break;

        default:
          tagName = 'noscript';
      }

      return tagName;
    }
  }, {
    key: 'render',
    value: function render() {
      var mjContent = this.props.mjContent;


      return _react2.default.createElement(this.getTagName(), {
        className: 'mj-raw',
        dangerouslySetInnerHTML: { __html: mjContent() }
      });
    }
  }]);

  return Raw;
}(_react.Component)) || _class);
exports.default = Raw;
