// mjml/lib/components/Section.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _cloneDeep2 = require('lodash/cloneDeep');

var _cloneDeep3 = _interopRequireDefault(_cloneDeep2);

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _MJMLElement = require('./decorators/MJMLElement');

var _MJMLElement2 = _interopRequireDefault(_MJMLElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Sections are intended to be used as rows within your email. They will be used to structure the layout.
 */
var Section = (_dec = (0, _MJMLElement2.default)({
  tagName: 'mj-section',
  attributes: {
    'background-repeat': 'repeat',
    'padding': '20px 0',
    'background-size': 'auto'
  }
}), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Section, _Component);

  function Section() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Section);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Section)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Section, [{
    key: 'isFullWidth',
    value: function isFullWidth() {
      var mjAttribute = this.props.mjAttribute;


      return mjAttribute('full-width') == 'full-width';
    }
  }, {
    key: 'getStyles',
    value: function getStyles() {
      var _props = this.props;
      var mjAttribute = _props.mjAttribute;
      var parentWidth = _props.parentWidth;


      var background = mjAttribute('background-url') ? {
        background: 'url(' + mjAttribute('background-url') + ') top center / ' + (mjAttribute('background-size') || '') + ' ' + (mjAttribute('background-repeat') || '')
      } : {
        background: mjAttribute('background-color')
      };

      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        td: {
          fontSize: 0,
          padding: mjAttribute('padding'),
          paddingBottom: mjAttribute('padding-bottom'),
          paddingLeft: mjAttribute('padding-left'),
          paddingRight: mjAttribute('padding-right'),
          paddingTop: mjAttribute('padding-top'),
          textAlign: mjAttribute('text-align'),
          verticalAlign: mjAttribute('vertical-align')
        },
        div: {
          maxWidth: parentWidth
        }
      }, {
        div: this.isFullWidth() ? {} : (0, _cloneDeep3.default)(background),
        table: this.isFullWidth() ? {} : (0, _cloneDeep3.default)(background),
        tableFullwidth: this.isFullWidth() ? (0, _cloneDeep3.default)(background) : {}
      });
    }
  }, {
    key: 'renderFullWidthSection',
    value: function renderFullWidthSection() {
      var mjAttribute = this.props.mjAttribute;


      return _react2.default.createElement(
        'table',
        {
          cellPadding: '0',
          cellSpacing: '0',
          'data-legacy-background': mjAttribute('background-url'),
          'data-legacy-border': '0',
          style: (0, _merge3.default)({}, this.styles.tableFullwidth, this.styles.table) },
        _react2.default.createElement(
          'tbody',
          null,
          _react2.default.createElement(
            'tr',
            null,
            _react2.default.createElement(
              'td',
              null,
              this.renderSection()
            )
          )
        )
      );
    }
  }, {
    key: 'renderSection',
    value: function renderSection() {
      var _props2 = this.props;
      var renderWrappedOutlookChildren = _props2.renderWrappedOutlookChildren;
      var mjAttribute = _props2.mjAttribute;
      var children = _props2.children;
      var parentWidth = _props2.parentWidth;

      var fullWidth = this.isFullWidth();

      return _react2.default.createElement(
        'div',
        { style: this.styles.div },
        _react2.default.createElement(
          'table',
          {
            cellPadding: '0',
            cellSpacing: '0',
            className: 'mj-section-outlook-background',
            'data-legacy-align': 'center',
            'data-legacy-background': fullWidth ? undefined : mjAttribute('background-url'),
            'data-legacy-border': '0',
            'data-url': mjAttribute('background-url') || '',
            'data-width': parentWidth,
            style: this.styles.table },
          _react2.default.createElement(
            'tbody',
            null,
            _react2.default.createElement(
              'tr',
              null,
              _react2.default.createElement(
                'td',
                { style: this.styles.td },
                renderWrappedOutlookChildren(children)
              )
            )
          )
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return this.isFullWidth() ? this.renderFullWidthSection() : this.renderSection();
    }
  }]);

  return Section;
}(_react.Component), _class2.baseStyles = {
  div: {
    margin: '0 auto'
  },
  table: {
    fontSize: 0,
    width: '100%'
  },
  td: {
    textAlign: 'center',
    verticalAlign: 'top'
  }
}, _temp2)) || _class);
exports.default = Section;
