// mjml/lib/components/Social.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _MJMLColumnElement = require('./decorators/MJMLColumnElement');

var _MJMLColumnElement2 = _interopRequireDefault(_MJMLColumnElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Social = (_dec = (0, _MJMLColumnElement2.default)({
  tagName: 'mj-social',
  attributes: {
    'facebook-content': 'Share',
    'facebook-href': '[[SHORT_PERMALINK]]',
    'facebook-icon-color': '#3b5998',
    'font-family': 'Ubuntu, Helvetica, Arial, sans-serif',
    'font-size': '13px',
    'google-content': '+1',
    'google-href': '[[SHORT_PERMALINK]]',
    'google-icon-color': '#dc4e41',
    'icon-size': '20px',
    'instagram-content': 'Share',
    'instagram-href': '[[SHORT_PERMALINK]]',
    'instagram-icon-color': '#3f729b',
    'line-height': '22px',
    'linkedin-content': 'Share',
    'linkedin-href': '[[SHORT_PERMALINK]]',
    'linkedin-icon-color': '#0077b5',
    'padding': '10px 25px',
    'mode': 'horizontal',
    'pinterest-content': 'Pin it',
    'pinterest-href': '[[SHORT_PERMALINK]]',
    'pinterest-icon-color': '#bd081c',
    'text-decoration': 'none',
    'text-mode': true,
    'twitter-content': 'Tweet',
    'twitter-href': '[[SHORT_PERMALINK]]',
    'twitter-icon-color': '#55acee',
    'align': 'center',
    'color': '#333333',
    'display': 'facebook:share twitter:share google:share',
    'base-url': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/'
  }
}), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Social, _Component);

  function Social() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    _classCallCheck(this, Social);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(Social)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Social, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return (0, _merge3.default)({}, this.constructor.baseStyles, {
        a: {
          color: mjAttribute('color'),
          fontFamily: mjAttribute('font-family'),
          fontSize: mjAttribute('font-size'),
          fontStyle: mjAttribute('font-style'),
          fontWeight: mjAttribute('font-weight'),
          lineHeight: mjAttribute('line-height'),
          textDecoration: mjAttribute('text-decoration')
        },
        td1: {
          padding: this.isHorizontal() ? '0 4px' : '4px 0'
        },
        td2: {
          width: mjAttribute('icon-size'),
          height: mjAttribute('icon-size')
        }
      });
    }
  }, {
    key: 'isHorizontal',
    value: function isHorizontal() {
      var mjAttribute = this.props.mjAttribute;


      return mjAttribute('mode') == 'horizontal';
    }
  }, {
    key: 'isInTextMode',
    value: function isInTextMode() {
      var mjAttribute = this.props.mjAttribute;


      return mjAttribute('text-mode') == true || mjAttribute('text-mode') == 'true';
    }
  }, {
    key: 'renderSocialButton',
    value: function renderSocialButton(platform, share) {
      var mjAttribute = this.props.mjAttribute;

      var definition = this.constructor.buttonDefinitions[platform];
      var href = share ? definition.linkAttribute.replace('[[URL]]', mjAttribute(platform + '-href')) : mjAttribute(platform + '-href');
      var iconStyle = {
        background: mjAttribute(platform + '-icon-color'),
        borderRadius: this.styles.img.borderRadius,
        width: mjAttribute('icon-size')
      };

      return _react2.default.createElement(
        'tr',
        { key: platform },
        _react2.default.createElement(
          'td',
          { style: this.styles.td1 },
          _react2.default.createElement(
            'table',
            {
              cellPadding: '0',
              cellSpacing: '0',
              'data-legacy-border': '0',
              style: iconStyle },
            _react2.default.createElement(
              'tbody',
              null,
              _react2.default.createElement(
                'tr',
                null,
                _react2.default.createElement(
                  'td',
                  { style: this.styles.td2 },
                  _react2.default.createElement(
                    'a',
                    { href: href },
                    _react2.default.createElement('img', {
                      alt: platform,
                      border: '0',
                      height: parseInt(mjAttribute('icon-size')),
                      src: mjAttribute('base-url') + definition.icon,
                      style: this.styles.img,
                      width: parseInt(mjAttribute('icon-size')) })
                  )
                )
              )
            )
          )
        ),
        this.isInTextMode() && _react2.default.createElement(
          'td',
          { style: this.styles.tdText },
          _react2.default.createElement('a', {
            dangerouslySetInnerHTML: { __html: mjAttribute(platform + '-content') },
            href: href,
            style: this.styles.a })
        )
      );
    }
  }, {
    key: 'renderSocialButtons',
    value: function renderSocialButtons() {
      var _this2 = this;

      var mjAttribute = this.props.mjAttribute;

      var platforms = mjAttribute('display');

      if (!platforms) {
        return;
      }

      return platforms.split(' ').map(function (label) {
        var platform = label.split(':')[0];
        var share = label.split(':')[1];
        if (!_this2.constructor.buttonDefinitions[platform]) {
          return;
        }

        return _this2.renderSocialButton(platform, share != "url");
      });
    }
  }, {
    key: 'renderHorizontal',
    value: function renderHorizontal() {
      var _this3 = this;

      var socialButtons = this.renderSocialButtons().map(function (socialButton, index) {
        return _react2.default.createElement(
          'table',
          {
            cellPadding: '0',
            cellSpacing: '0',
            'data-legacy-align': 'left',
            'data-legacy-border': '0',
            key: 'wrapped-social-button-' + index,
            style: _this3.styles.tableHorizontal },
          _react2.default.createElement(
            'tbody',
            null,
            socialButton
          )
        );
      }).reduce(function (result, socialButton, index) {
        result.push(socialButton);
        result.push(_react2.default.createElement('div', { className: 'mj-social-outlook-line', key: 'outlook-line-' + index }));

        return result;
      }, [_react2.default.createElement('div', { className: 'mj-social-outlook-open', key: 'outlook-open' })]);

      socialButtons[socialButtons.length - 1] = _react2.default.createElement('div', { className: 'mj-social-outlook-close', key: 'outlook-close' });

      return socialButtons;
    }
  }, {
    key: 'renderVertical',
    value: function renderVertical() {
      return _react2.default.createElement(
        'table',
        {
          cellPadding: '0',
          cellSpacing: '0',
          'data-legacy-align': 'center',
          'data-legacy-border': '0',
          style: this.styles.tableVertical },
        _react2.default.createElement(
          'tbody',
          null,
          this.renderSocialButtons()
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { style: this.styles.div },
        this.isHorizontal() ? this.renderHorizontal() : this.renderVertical()
      );
    }
  }]);

  return Social;
}(_react.Component), _class2.baseStyles = {
  div: {
    textAlign: 'center'
  },
  tableHorizontal: {
    float: 'none',
    display: 'inline-table'
  },
  tableVertical: {
    margin: '0 auto'
  },
  td1: {
    verticalAlign: 'middle'
  },
  td2: {
    textAlign: 'center',
    verticalAlign: 'middle'
  },
  tdText: {
    padding: '8px 8px 8px 0',
    verticalAlign: 'middle'
  },
  a: {
    textDecoration: 'none',
    display: 'block',
    borderRadius: '3px'
  },
  img: {
    display: 'block',
    borderRadius: '3px'
  }
}, _class2.buttonDefinitions = {
  facebook: {
    linkAttribute: 'https://www.facebook.com/sharer/sharer.php?u=[[URL]]',
    icon: 'facebook.png',
    textModeContent: 'Share'
  },
  twitter: {
    linkAttribute: 'https://twitter.com/home?status=[[URL]]',
    icon: 'twitter.png',
    textModeContent: 'Tweet'
  },
  google: {
    linkAttribute: 'https://plus.google.com/share?url=[[URL]]',
    icon: 'google-plus.png',
    textModeContent: '+1'
  },
  pinterest: {
    linkAttribute: 'https://pinterest.com/pin/create/button/?url=[[URL]]&ampmedia=&ampdescription=',
    icon: 'pinterest.png',
    textModeContent: 'Pin it'
  },
  linkedin: {
    linkAttribute: 'https://www.linkedin.com/shareArticle?mini=true&ampurl=[[URL]]&amptitle=&ampsummary=&ampsource=',
    icon: 'linkedin.png',
    textModeContent: 'Share'
  },
  instagram: {
    linkAttribute: '[[URL]]',
    icon: 'instagram.png',
    textModeContent: 'Share'
  }
}, _temp2)) || _class);
exports.default = Social;
