// mjml/lib/cli.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initComponent = exports.watch = exports.renderStream = exports.renderFile = exports.version = undefined;

var _index = require('./index');

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var capitalize = function capitalize(name) {
  return name.charAt(0).toUpperCase() + name.slice(1).toLowerCase().replace(/-/g, '');
};

/*
 * The version number is the NPM
 * version number. It should be the same as the MJML engine
 */
var version = exports.version = _index.version;

/*
 * Turns a callback style to a Promise style one
 */
var promisify = function promisify(fn) {
  return function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return new Promise(function (resolve, reject) {
      return fn.apply(undefined, _toConsumableArray(args.concat(function (err) {
        for (var _len2 = arguments.length, data = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
          data[_key2 - 1] = arguments[_key2];
        }

        return err ? reject(err) : resolve.apply(undefined, data);
      })));
    });
  };
};

/*
 * Minimal Error Handling
 */
var error = function error(e) {
  return console.log(e.stack ? e.stack : e);
};

/*
 * Stdin to string buffer
 */
var stdinToBuffer = function stdinToBuffer(stream, callback) {
  var buffer = '';

  stream.on('data', function (chunck) {
    buffer += chunck;
  });

  stream.on('end', function () {
    callback(null, buffer);
  });
};

/*
 * Utility functions
 * write: write to a file
 * read: read a fileexists: ensure the file exists
 */
var write = promisify(_fs2.default.writeFile);
var read = promisify(_fs2.default.readFile);
var exists = promisify(function (file, cb) {
  return _fs2.default.access(file, _fs2.default.R_OK | _fs2.default.W_OK, cb);
});
var readStdin = promisify(stdinToBuffer);

/*
 * Render an input promise
 */
var render = function render(bufferPromise, _ref) {
  var min = _ref.min;
  var output = _ref.output;
  var stdout = _ref.stdout;

  bufferPromise.then(function (mjml) {
    return (0, _index.mjml2html)(mjml.toString(), { minify: min });
  }).then(function (result) {
    return stdout ? process.stdout.write(result) : write(output, result);
  }).catch(error);
};

/*
 * Turns an MJML input file into a pretty HTML file
 * min: boolean that specify the output format (pretty/minified)
 */
var renderFile = exports.renderFile = function renderFile(input, options) {
  return render(exists(input).then(function () {
    return read(input);
  }), options);
};

/**
 * Render based on input stream
 */
var renderStream = exports.renderStream = function renderStream(options) {
  return render(readStdin(process.stdin), options);
};

/*
 * Watch changes on a specific input file by calling render on each change
 */
var watch = exports.watch = function watch(input, options) {
  return _fs2.default.watchFile(input, function () {
    return renderFile(input, options);
  });
};

/*
* Return the code of an MJML component for a given name
*/
var createComponent = function createComponent(name, ending) {
  var lowerName = name.toLowerCase();

  return '\nimport { MJMLColumnElement, elements, registerElement } from \'mjml\'\nimport merge from \'lodash/merge\'\nimport React, { Component } from \'react\'\n\n/*\n * Wrap your dependencies here.\n */\nconst { text: MjText } = elements\n\nconst NAME = \'' + lowerName + '\'\n\n@MJMLColumnElement({\n  tagName: \'mj-' + lowerName + '\',\n  content: \' \',\n\n  /*\n   * These are your default css attributes\n   */\n  attributes: {\n    \'color\': \'#424242\',\n    \'font-family\': \'Helvetica\',\n    \'margin-top\': \'10px\'\n  }\n})\nclass ' + name + ' extends Component {\n\n  /*\n   * Build your styling here\n   */\n  getStyles () {\n    const { mjAttribute, color } = this.props\n\n    return merge({}, this.constructor.baseStyles, {\n      text: {\n      /*\n       * Get the color attribute\n       * Example: <mj-' + lowerName + ' color="blue">content</mj-' + lowerName + '>\n       */\n        color: mjAttribute(\'color\')\n      }\n    })\n  }\n\n  render () {\n    const css = this.getStyles()\n    const content = \'Hello World!\'\n\n    return (\n      <MjText style={ css }>\n        { content }\n      </MjText>\n    )\n  }\n\n}\n\nregisterElement(\'' + lowerName + '\', ' + name + (ending ? ', { endingTag: true }' : '') + ')\n\nexport default ' + name + '\n';
};

/*
 * Create a new component based on the default template
 */
var initComponent = exports.initComponent = function initComponent(name, ending) {
  return write('./' + capitalize(name) + '.js', createComponent(capitalize(name), ending)).then(function () {
    return console.log('Component created: ' + capitalize(name));
  });
};
