// mjml/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.version = exports.registerElement = exports.elements = exports.MJMLElement = exports.MJMLColumnElement = exports.mjml2html = exports.documentParser = undefined;

var _MJMLElementsCollection = require('./MJMLElementsCollection');

Object.defineProperty(exports, 'registerElement', {
  enumerable: true,
  get: function get() {
    return _MJMLElementsCollection.registerElement;
  }
});

var _documentParser2 = require('./documentParser');

var _documentParser3 = _interopRequireDefault(_documentParser2);

var _mjml2html2 = require('./mjml2html');

var _mjml2html3 = _interopRequireDefault(_mjml2html2);

var _MJMLColumnElement2 = require('./components/decorators/MJMLColumnElement');

var _MJMLColumnElement3 = _interopRequireDefault(_MJMLColumnElement2);

var _MJMLElement2 = require('./components/decorators/MJMLElement');

var _MJMLElement3 = _interopRequireDefault(_MJMLElement2);

var _MJMLElementsCollection2 = _interopRequireDefault(_MJMLElementsCollection);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.documentParser = _documentParser3.default;
exports.mjml2html = _mjml2html3.default;
exports.MJMLColumnElement = _MJMLColumnElement3.default;
exports.MJMLElement = _MJMLElement3.default;
exports.elements = _MJMLElementsCollection2.default;
var version = exports.version = function version() {
  return '1.3.4';
};
