// mjml/lib/components/decorators/MJMLColumnElement.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _merge2 = require('lodash/merge');

var _merge3 = _interopRequireDefault(_merge2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _hoistNonReactStatics = require('hoist-non-react-statics');

var _hoistNonReactStatics2 = _interopRequireDefault(_hoistNonReactStatics);

var _MJMLElement = require('./MJMLElement');

var _MJMLElement2 = _interopRequireDefault(_MJMLElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function createComponent(ComposedComponent, defaultMJMLDefinition) {
  var _dec, _class, _class2, _temp2;

  var MJMLColumnElement = (_dec = (0, _MJMLElement2.default)(defaultMJMLDefinition), _dec(_class = (_temp2 = _class2 = function (_Component) {
    _inherits(MJMLColumnElement, _Component);

    function MJMLColumnElement() {
      var _Object$getPrototypeO;

      var _temp, _this, _ret;

      _classCallCheck(this, MJMLColumnElement);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(MJMLColumnElement)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(MJMLColumnElement, [{
      key: 'getStyles',
      value: function getStyles() {
        var mjAttribute = this.props.mjAttribute;


        return (0, _merge3.default)({}, this.constructor.baseStyles, {
          td: {
            background: mjAttribute('container-background-color'),
            fontSize: 0,
            padding: mjAttribute('padding'),
            paddingTop: mjAttribute('padding-top'),
            paddingBottom: mjAttribute('padding-bottom'),
            paddingRight: mjAttribute('padding-right'),
            paddingLeft: mjAttribute('padding-left')
          }
        });
      }
    }, {
      key: 'render',
      value: function render() {
        var mjAttribute = this.props.mjAttribute;


        return _react2.default.createElement(
          'tr',
          null,
          _react2.default.createElement(
            'td',
            {
              'data-legacy-align': mjAttribute('align'),
              'data-legacy-background': mjAttribute('container-background-color'),
              style: this.styles.td },
            _react2.default.createElement(ComposedComponent, this.props)
          )
        );
      }
    }]);

    return MJMLColumnElement;
  }(_react.Component), _class2.baseStyles = {
    td: {
      wordBreak: 'break-word'
    }
  }, _temp2)) || _class);


  (0, _hoistNonReactStatics2.default)(MJMLColumnElement, ComposedComponent);

  return MJMLColumnElement;
}

exports.default = function (defaultMJMLDefinition) {
  if (typeof defaultMJMLDefinition == 'function') {
    return createComponent(defaultMJMLDefinition);
  }

  return function (ComposedComponent) {
    return createComponent(ComposedComponent, defaultMJMLDefinition);
  };
};
